### PR TITLE
Add support for executing some operations in batch mode

### PR DESCRIPTION
**Pull request by aclemons**

### CODE REVIEW

**Comment by aclemons** [comment]
Keen to start getting feedback on this.

**Comment by aclemons** [comment]
Any update on this?

**Comment by rg3** [comment]
First off, sorry for the long time to reply and the following wall of text, and thank you for your patience. I had completely forgotten about this until you asked for feedback again some days ago.

Your approach to batch mode is pretty simple and uninstrusive, so I'd like to merge the functionality if you find it useful, but I'd like many minor changes to take place. I could do this myself, but I prefer if you change it yourself as if you were the maintainer. Here's the list and don't hesitate to ask anything or discuss any item.

Batch mode global flag: good and simple idea.

Help files: as you can prepend the batch operation to any other operation, I'd like to see the operation added to the full list in "all.txt" under "Other operations" and have its own "batch.txt" help file instead of modifying other files because there are too many of them affected, in my opinion. For example, info, install or reinstall do not mention the new batch mode and they're affected.

The help file for the batch operation would mention the essentials: another operation follows and the program will error out if user input would be needed to select a specific package version among several ones. In addition, all other interactive operations are done automatically with a predefined and documented behavior.

There are two blank lines before SlackrollBatchModeError instead of one. It could inherit from SlackrollError.

What's the idea behind the choose_option() changes? I get the impression it's about printing what slackroll is doing as if users were making the choices themselves instead of being silent, but I find it's too verbose when doing that. For example, in handle_dotnew_files_installation it chooses a default option that varies with the situation. Instead of the current approach, I'd create two versions: handle_dotnew_files_installation_interactive and handle_dotnew_files_installation_batch. handle_dotnew_files_installation would call either one depending on the batch mode global flag.

The batch version would be short and simple, iterating over the list, getting the old file name with old_file() like the current version does and deciding what to do, printing a single line per action. It's important the lines look different at first sight depending on the action taken ("Keeping both /foo and /foo.new for manual review", "Renaming /foo.new to /foo because /foo does not exist", "Ignoring nonexistent /foo.new and keeping /foo", "Warning: neither /foo.new nor /foo exist"). It would be a very short and explicit function with tight output suitable to be read if cron mails you the output.

Then I'd do the same in handle_dotnew_files_removal(). I understand you modify less lines with your current approach, but the extra lines would make the code more explicit.

About cron output: if cron mails you the output, maybe we don't want to print the download progress as usual in batch mode. Should we silence other parts of the output? Do cron emails look sane now or do you redirect output to a file to be reviewed if the batch operation fails?

confirm_continue() is very short so I'd remove the blank line. It could be called maybe_confirm_continue() because it now depends on the batch mode flag.

post_kernel_operation() could be made similar to confirm_continue(): detect if we're in batch mode and simply do nothing instead of printing the menu and choosing to drop out automatically. It could print a message before the early exit, warning about the possible need to update the boot loader configuration file, rebooting, etc. If I were to use batch mode I don't know if I'd run the kernel-upgrade and kernel-clean operations. Do you run them? What's the sanest way of handling it?

The *_interactive() and *_batch() approach I suggest screams for object-oriented code or similar: a class that implements UI-related methods with a subclass for interactive mode and another one for batch mode. This way, the decision about which set of operations to use can be made in a single instantiation line. However, with the current state of the code it's simply not worth it, in my opinion.

"batch" needs to be added to op_num_args with an argument count of -1 so its arguments can be verified and so it enters the suggested command list if it's mistyped (let's say as "bathc"). At the same time, the operation after "batch" needs to be verified again and cannot be "batch". I think I'd put the whole verification block ("# Verify number of arguments") in its own function, receiving the operations map (op_num_args), operation and arguments array as function arguments, and I would call it twice. First, with the whole list including "batch". Once verified, if it's "batch", I'd activate the batch mode flag, remove "batch" from op_num_args (del op_num_args['batch']), reassign operation and args like you do now and call the verification function again. This new function could be called "verify_operation_and_args".

Finally, the gh-pages branch contains the website files. There's a section called "feedback" that references batch mode. The text should be changed.

Please give me your feedback on this. When batch mode is ready, we can release version 47. This implies changing the version number in slackroll_version, generating the package, uploading it to GitHub, changing the URL and version in gh-pages and adding a new item in the RSS file (in gh-pages).

**Comment by aclemons** [comment]
I'll have a look at each of these and post back here with updates/changes. Thanks for the thorough review.

**Comment by aclemons** [comment]
Maybe you could go ahead and cherry-pick fa7beca (whitespace fixups) into master so we can keep this PR nice and clearly arranged?

**Comment by rg3** [comment]
Sure. Done.

**Comment by aclemons** [comment]
> If I were to use batch mode I don't know if I'd run the kernel-upgrade and kernel-clean operations. Do you run them? What's the sanest way of handling it?

I'm calling slackroll in batch mode from other scripts. There I have some extra logic around what things are upgraded (inspecting the transient list). When kernel upgrades happen, I use kernel-upgrade to install the new one. The script then handles regenerating the initrd and grub config. Sure, we could add some option to indicate a script to be called in batch mode for the post_kernel_operation, but I think the sanest thing is to do nothing and assume whatever is calling slackroll in batch mode knows it needs to handle any kernel things itself. I don't actually allow the batch upgrading of kernels to happen from a cron job. When a kernel upgrade is released, I run the same script with a flag to allow kernel upgrades and then do the reboots etc making sure everything is still working.

**Comment by aclemons** [comment]
> There are two blank lines before SlackrollBatchModeError instead of one. It could inherit from SlackrollError.

done

**Comment by aclemons** [comment]
> What's the idea behind the choose_option() changes?

There was no planning on my part here. It just seemed like the least intrusive way to get it working for me. I wasn't sure if there was a chance to get batch mode merged, so I implemented the simplest, least intrusive change I could.

Your suggestion seems reasonable so I'll update the PR for that.

**Comment by aclemons** [comment]
> The batch version would be short and simple, iterating over the list
> [...]

I've updated this now. I like this much better than what I had previously.

**Comment by aclemons** [comment]
> confirm_continue() is very short so I'd remove the blank line. It could be called maybe_confirm_continue() because it now depends on the batch mode flag.

Done.

**Comment by aclemons** [comment]
> post_kernel_operation() could be made similar to confirm_continue():
> [...]

Done

**Comment by aclemons** [comment]
> Should we silence other parts of the output?

Yes, I think so. The script I run as a cron job usually does an update, then clean-cache, then download-upgrades. I'm currently redirecting the output of the last one to /dev/null for the reasons you mentioned.

I have changed slackroll to not print the download progress now when in batch mode.

**Comment by aclemons** [comment]
> The *_interactive() and *_batch() approach I suggest screams for object-oriented code  [...] However, with the current state of the code it's simply not worth it, in my opinion.

Yes, the script is very procedural at the moment so it would not fit that well. I'm actually a ruby developer and python is not my specialty, but it would be nice to have objects with spec coverage for all the operations. Something for the future I guess.

**Comment by aclemons** [comment]
I'll look to update the docs and the arg count validation tomorrow.

**Comment by aclemons** [comment]
> batch needs to be added to op_num_args with an argument count[...]

Have added this now. Much nicer.

**Comment by aclemons** [comment]
> Help files [...]

Ok I have moved these around. Not sure if there is enough in batch.txt or if you are keen for more elaboration.

**Comment by aclemons** [comment]
That should be it. Ready for re-review now.

**Comment by aclemons** [comment]
Also removed the text about non-interactive usage being unsupported in the gh-pages branch in my fork. We can add the release notes etc after merging and final testing.

**Comment by aclemons** [comment]
We also probably want to disable output pagination in batch mode.

**Comment by aclemons** [comment]
The pagination only happens if stdout is a tty, but I was just running some commands manually with batch mode enabled and noticed the pagination. That is probably not expected. Thoughts?

**Comment by rg3** [comment]
Yes, pagination should be disabled in batch mode. If you use batch mode in a terminal it should not require user interaction. Pagination is user interaction, so batch mode disables pagination too.

I'll review the rest of the changes today.

File: helpfiles/all.txt
Comment (rg3) [inline helpfiles/all.txt]: Put "ARGUMENT..." inside brackets to hint about arguments being optional depending on the operation.

**Comment by rg3** [comment]
We're much closer now to getting this merged. I started a review on the code. I hope it's easier for you to handle my comments that way.

File: helpfiles/batch.txt
Comment (rg3) [inline helpfiles/batch.txt]: The same about "ARGUMENT..."

File: helpfiles/batch.txt
Comment (rg3) [inline helpfiles/batch.txt]: Remove "and documented". It's implicit because the documentation follows.

File: helpfiles/batch.txt
Comment (rg3) [inline helpfiles/batch.txt]: Remove the list header and simply include the following points, literally or reworded.

- On package removal, every .new file pair left behind will be removed.
- On package installation, if a single file exists in a .new file pair, it will be kept without the .new suffix. If both files exist, both files will be kept for manual review later.
- No interactive confirmation prompts will be present.
- No pagination will be used.

File: slackroll
Comment (rg3) [inline slackroll]: Put this "if" block before creating the options list above, which would not be used.

File: slackroll
Comment (rg3) [inline slackroll:1165]: I'd personally remove the blank lines in this if/else block but it's a matter of style.

File: slackroll
Comment (rg3) [inline slackroll]: Fine but I'd word this more carefully, using "may" instead of "will". For example:

Warning: you may need to modify your bootloader configuration and reboot.

File: slackroll
Comment (rg3) [inline slackroll]: Again, I'd probably remove some whitespace here.

File: slackroll
Comment (rg3) [inline slackroll]: I'd move the op_num_args argument to be the first or the last one, so as to keep "operation" and "args" next to each other. I made a mistake when choosing the argument order in verify_num_args(), which you probably took as a reference. That could be changed too.

File: slackroll
Comment (rg3) [inline slackroll]: You're not removing "batch" before the second verification, so "slackroll batch batch batch", "slackroll batch batch list-transient", etc, are legal now and do nothing. Use "del op_num_args['batch']" before the second verification to forbid it.

**Review by rg3 — CHANGES_REQUESTED** [review]

File: helpfiles/all.txt
Comment (aclemons) [reply helpfiles/all.txt]: Addressed

File: helpfiles/batch.txt
Comment (aclemons) [reply helpfiles/batch.txt]: Addressed

File: helpfiles/batch.txt
Comment (aclemons) [reply helpfiles/batch.txt]: Addressed

File: helpfiles/batch.txt
Comment (aclemons) [reply helpfiles/batch.txt]: I opted to include your texts here verbatim.

File: slackroll
Comment (aclemons) [reply slackroll]: Addressed

File: slackroll
Comment (aclemons) [reply slackroll:1165]: Yes, sorry, ruby style :)

File: slackroll
Comment (aclemons) [reply slackroll]: Updated

File: slackroll
Comment (aclemons) [reply slackroll]: Updated

File: slackroll
Comment (aclemons) [reply slackroll]: Ok, I've updated the signatures and usages of both of those methods.

File: slackroll
Comment (aclemons) [reply slackroll]: Ah yes nice catch. Updated now.

**Comment by aclemons** [comment]
I've pushed this up into a commit, will squash once it's all checked off.

**Comment by aclemons** [comment]
For the pagination:

Is it ok to just access the global var like this:

```
$ git diff
diff --git a/slackroll b/slackroll
index 1422cd2..b095a70 100755
--- a/slackroll
+++ b/slackroll
@@ -711,7 +711,7 @@ class SlackrollOutputInterceptor(object): # Intercepts stdout and uses a pager i
         output = self.__buffer.getvalue()
         self.__buffer.close()
         numlines = output.count('\n')
-        if too_many_lines(numlines):
+        if not slackroll_batch_mode and too_many_lines(numlines):
             proc = call_pager()
             proc.stdin.write(output)
             proc.stdin.close()
```

**Comment by rg3** [comment]
It's OK to use slackroll_batch_mode there, but I'd rename too_many_lines() to needs_pager() and put every condition in there to keep them together.

`if (not sys.stdout.isatty()) or slackroll_batch_mode:`

OK to everything else.

**Comment by aclemons** [comment]
Ok, updated now.

**Comment by rg3** [comment]
Merged. I fixed some additional last-minute problems I found while reviewing the code again. Minor problems.

**Pull request merged**
